lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -0,0 +1,243 @@
+
+      // Record all the RefType ports to be removed later.
+      for (size_t portNum = 0, e = module.getNumPorts(); portNum < e; ++portNum)
+        if (module.getPortType(portNum).isa<RefType>())
+          refPortsToRemoveMap[module].push_back(portNum);
+    }
+
+    // Now erase all the Ops and ports of RefType.
+    // This needs to be done as the last step to ensure uses are erased before
+    // the def is erased.
+    for (auto op : llvm::reverse(opsToRemove))
+      op->erase();
+    for (auto iter : refPortsToRemoveMap)
+      if (auto mod = dyn_cast<FModuleOp>(iter.getFirst()))
+        mod.erasePorts(iter.getSecond());
+      else if (auto inst = dyn_cast<InstanceOp>(iter.getFirst())) {
+        ImplicitLocOpBuilder b(inst.getLoc(), inst);
+        inst.erasePorts(b, iter.getSecond());
+        inst.erase();
+      }
+  }
+
+  ArrayAttr getRemoteRefSend(Value val, Operation *op) {
+    auto iter = reachingRefSendAt.find(val);
+    if (iter != reachingRefSendAt.end())
+      return iter->getSecond();
+    // The referenced module must have already been analyzed, error out if the
+    // dataflow at the child module is not resolved.
+    op->emitOpError(
+        "reference dataflow cannot be traced back to the remote read op");
+    return {};
+  }
+  //
+  // Replace the RefResolveOp with verbatim op representing the XMR.
+  LogicalResult handleRefResolve(RefResolveOp resolve) {
+    opsToRemove.push_back(resolve);
+    auto remoteOpPath = getRemoteRefSend(resolve.getRef(), resolve);
+    if (!remoteOpPath)
+      return failure();
+    auto xmrSize = remoteOpPath.size();
+    SmallVector<Attribute> xmrHierPath(xmrSize);
+    SmallString<128> xmrString;
+    for (auto instanceRef : llvm::enumerate(remoteOpPath)) {
+      xmrHierPath[xmrSize - instanceRef.index() - 1] = instanceRef.value();
+      ("{{" + Twine(instanceRef.index()) + "}}").toVector(xmrString);
+      if (instanceRef.index() < xmrSize - 1)
+        xmrString += '.';
+    }
+
+    // The source of the dataflow for this RefResolveOp is established. So
+    // replace the RefResolveOp with the coresponding VerbatimExpr to
+    // generate the XMR.
+    ImplicitLocOpBuilder builder(resolve.getLoc(), resolve);
+    auto xmrVerbatim =
+        builder.create<VerbatimExprOp>(resolve.getType().cast<FIRRTLType>(),
+                                       xmrString, ValueRange{}, xmrHierPath);
+    resolve.getResult().replaceAllUsesWith(xmrVerbatim);
+    return success();
+  }
+
+  // Propagate the reachable RefSendOp across modules.
+  LogicalResult handleInstanceOp(InstanceOp inst) {
+    auto refMod = dyn_cast<FModuleOp>(inst.getReferencedModule());
+    for (size_t portNum = 0, e = inst.getNumResults(); portNum < e; ++portNum) {
+      auto instanceResult = inst.getResult(portNum);
+      if (!instanceResult.getType().isa<RefType>())
+        continue;
+      if (!refMod) {
+        return inst.emitOpError("cannot lower ext modules with RefType ports");
+      }
```

Comment:
Nit:
```suggestion
      if (!refMod) 
        return inst.emitOpError("cannot lower ext modules with RefType ports");
```

Great to add this check. We should, going forward, think about how to handle circuits that do this.